### ISSUE TITLE
Let frames/coordinates have (internal?) has_distance function

### ISSUE THREAD

**Issue by mhvk** [astropy/astropy#10294]
In a fair number of places inside our coordinate machinery, we have to deal with coordinates with and without distances, and are testing this somewhat haphazardly - checking whether the representation is `UnitSphericalRepresentation` *or* whether it is `Spherical` and the distance is all unity. We should normalize these tests, and ideally also ensure that the second case simply doesn't happen (so that the test can be very simple).

**Comment by CaioCoutinhoP** [comment]
This issue is still open for contributions?

**Comment by mhvk** [comment]
Yes, it has not been tackled yet. It does require fairly detailed understanding of how about coordinate machinery works, though!
